tictactoe/io.py

```diff
@@ -0,0 +1,75 @@
+from abc import ABC, abstractmethod
+from tictactoe import Cell
+
+import click
+from terminaltables import SingleTable
+
+
+class IOFrontend(ABC):
+    @abstractmethod
+    def print_board(self, board):
+        pass
+
+    @abstractmethod
+    def print_winner(self, name=None):
+        pass
+
+    @abstractmethod
+    def get_input(self):
+        pass
+
+
+class ConsoleFrontend(IOFrontend):
+    placeholders = ["①", "②", "③", "④", "⑤", "⑥", "⑦", "⑧", "⑨"]
```

Comment:
I'm concerned that we don't have a basic frontend that will work on everyone's terminal. Not all terminals will display emojis. I think it would be really nice to have a simple frontend that just worked for everyone, then a fancier frontend (like the existing `ConsoleFrontend`) to illustrate how to plug in different frontends.

Here's what I'm thinking as a super simple frontend:

```python
class BasicConsoleFrontend(IOFrontend):
    def print_board(self, board):
        for i, row in enumerate(board):
            for j, column in enumerate(row):
                if cell != Cell.EMPTY:
                    print(cell, end = "|")
                else:
                    cell_number = i * len(row) + j + 1
                    print(cell_number)
            print()
        print()

    def print_winner(self, name=None):
        if name is None:
            print("Game is a draw!")
        else:
            print(f"Player {name} wins!")

    def get_input(self):
        return input("Enter a number of the cell: ")
```

You could also simplify the nested loop a bit:

```python
for i, row in enumerate(board):
    row_display = []
    for j, col in enumerate(row):
        if cell != Cell.EMPTY:
            row_display.append[cell]
        else:
            cell_number = i * len(row) + j + 1
            row_display.append[cell_number]
    print(" | ".join(row_display)
```

That changes the output a bit so that there isn't the | on the far right-hand side of the table, and avoids the two calls to `print()` without any arguments.

So lots of food for thought there. Curious to know what you think!

Reply:
I believe every modern terminal/shell should support proper unicode. Anyway agree to simplify this to just ASCII. The point of doing that initially was a comment from @gahjelle to have some nicer UI.

Reply:
![Screen Shot 2020-08-25 at 10 57 12 AM](https://user-images.githubusercontent.com/3602533/91148039-c1a34200-e6c1-11ea-88f7-49e55b955863.png)

Reply:
@gahjelle Does love his pretty terminal UIs 🙂

> I believe every modern terminal/shell should support proper Unicode.

Not all readers use "modern" terminals, though. I think having ASCII as a default is the best option. It still looks good and is pretty much guaranteed to work anywhere.